From patchwork Fri Feb 22 14:38:58 2019
Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: 7bit
X-Patchwork-Submitter: Stefan Raspl <raspl@linux.ibm.com>
X-Patchwork-Id: 10826203
Return-Path: <kvm-owner@kernel.org>
Received: from mail.wl.linuxfoundation.org (pdx-wl-mail.web.codeaurora.org
 [172.30.200.125])
	by pdx-korg-patchwork-2.web.codeaurora.org (Postfix) with ESMTP id 5AC47180E
	for <patchwork-kvm@patchwork.kernel.org>;
 Fri, 22 Feb 2019 14:39:12 +0000 (UTC)
Received: from mail.wl.linuxfoundation.org (localhost [127.0.0.1])
	by mail.wl.linuxfoundation.org (Postfix) with ESMTP id 47469305E7
	for <patchwork-kvm@patchwork.kernel.org>;
 Fri, 22 Feb 2019 14:39:12 +0000 (UTC)
Received: by mail.wl.linuxfoundation.org (Postfix, from userid 486)
	id 3974330731; Fri, 22 Feb 2019 14:39:12 +0000 (UTC)
X-Spam-Checker-Version: SpamAssassin 3.3.1 (2010-03-16) on
	pdx-wl-mail.web.codeaurora.org
X-Spam-Level: 
X-Spam-Status: No, score=-7.9 required=2.0 tests=BAYES_00,MAILING_LIST_MULTI,
	RCVD_IN_DNSWL_HI autolearn=ham version=3.3.1
Received: from vger.kernel.org (vger.kernel.org [209.132.180.67])
	by mail.wl.linuxfoundation.org (Postfix) with ESMTP id AB0AB305E7
	for <patchwork-kvm@patchwork.kernel.org>;
 Fri, 22 Feb 2019 14:39:11 +0000 (UTC)
Received: (majordomo@vger.kernel.org) by vger.kernel.org via listexpand
        id S1726560AbfBVOjL (ORCPT
        <rfc822;patchwork-kvm@patchwork.kernel.org>);
        Fri, 22 Feb 2019 09:39:11 -0500
Received: from mx0a-001b2d01.pphosted.com ([148.163.156.1]:50732 "EHLO
        mx0a-001b2d01.pphosted.com" rhost-flags-OK-OK-OK-OK)
        by vger.kernel.org with ESMTP id S1725978AbfBVOjK (ORCPT
        <rfc822;kvm@vger.kernel.org>); Fri, 22 Feb 2019 09:39:10 -0500
Received: from pps.filterd (m0098396.ppops.net [127.0.0.1])
        by mx0a-001b2d01.pphosted.com (8.16.0.27/8.16.0.27) with SMTP id
 x1MEYq1W115902
        for <kvm@vger.kernel.org>; Fri, 22 Feb 2019 09:39:09 -0500
Received: from e06smtp05.uk.ibm.com (e06smtp05.uk.ibm.com [195.75.94.101])
        by mx0a-001b2d01.pphosted.com with ESMTP id 2qtj20kkx1-1
        (version=TLSv1.2 cipher=AES256-GCM-SHA384 bits=256 verify=NOT)
        for <kvm@vger.kernel.org>; Fri, 22 Feb 2019 09:39:09 -0500
Received: from localhost
        by e06smtp05.uk.ibm.com with IBM ESMTP SMTP Gateway: Authorized Use
 Only! Violators will be prosecuted
        for <kvm@vger.kernel.org> from <raspl@linux.ibm.com>;
        Fri, 22 Feb 2019 14:39:04 -0000
Received: from b06cxnps3075.portsmouth.uk.ibm.com (9.149.109.195)
        by e06smtp05.uk.ibm.com (192.168.101.135) with IBM ESMTP SMTP Gateway:
 Authorized Use Only! Violators will be prosecuted;
        (version=TLSv1/SSLv3 cipher=AES256-GCM-SHA384 bits=256/256)
        Fri, 22 Feb 2019 14:39:02 -0000
Received: from d06av24.portsmouth.uk.ibm.com (mk.ibm.com [9.149.105.60])
        by b06cxnps3075.portsmouth.uk.ibm.com (8.14.9/8.14.9/NCO v10.0) with
 ESMTP id x1MEd1E162914724
        (version=TLSv1/SSLv3 cipher=DHE-RSA-AES256-GCM-SHA384 bits=256
 verify=FAIL);
        Fri, 22 Feb 2019 14:39:01 GMT
Received: from d06av24.portsmouth.uk.ibm.com (unknown [127.0.0.1])
        by IMSVA (Postfix) with ESMTP id 68E9D42042;
        Fri, 22 Feb 2019 14:39:01 +0000 (GMT)
Received: from d06av24.portsmouth.uk.ibm.com (unknown [127.0.0.1])
        by IMSVA (Postfix) with ESMTP id 58AD44203F;
        Fri, 22 Feb 2019 14:39:01 +0000 (GMT)
Received: from tuxmaker.boeblingen.de.ibm.com (unknown [9.152.85.9])
        by d06av24.portsmouth.uk.ibm.com (Postfix) with ESMTPS;
        Fri, 22 Feb 2019 14:39:01 +0000 (GMT)
Received: from tuxmaker.boeblingen.de.ibm.com (localhost [127.0.0.1])
        by tuxmaker.boeblingen.de.ibm.com (Postfix) with ESMTP id 02BBF20F5F6;
        Fri, 22 Feb 2019 15:39:01 +0100 (CET)
From: Stefan Raspl <raspl@linux.ibm.com>
To: kvm@vger.kernel.org
Cc: drjones@redhat.com, pbonzini@redhat.com, sergeyb@bronevichok.ru,
        heiko.carstens@de.ibm.com
Subject: [kvm-unit-tests PATCH v2 1/2] tap13: list testcases individually
Date: Fri, 22 Feb 2019 15:38:58 +0100
X-Mailer: git-send-email 2.16.4
In-Reply-To: <20190222143859.15382-1-raspl@linux.ibm.com>
References: <20190222143859.15382-1-raspl@linux.ibm.com>
X-TM-AS-GCONF: 00
x-cbid: 19022214-0020-0000-0000-00000319FF2F
X-IBM-AV-DETECTION: SAVI=unused REMOTE=unused XFE=unused
x-cbparentid: 19022214-0021-0000-0000-0000216B57A0
Message-Id: <20190222143859.15382-2-raspl@linux.ibm.com>
X-Proofpoint-Virus-Version: vendor=fsecure engine=2.50.10434:,,
 definitions=2019-02-22_10:,,
 signatures=0
X-Proofpoint-Spam-Details: rule=outbound_notspam policy=outbound score=0
 priorityscore=1501
 malwarescore=0 suspectscore=1 phishscore=0 bulkscore=0 spamscore=0
 clxscore=1015 lowpriorityscore=0 mlxscore=0 impostorscore=0
 mlxlogscore=999 adultscore=0 classifier=spam adjust=0 reason=mlx
 scancount=1 engine=8.0.1-1810050000 definitions=main-1902220104
Sender: kvm-owner@vger.kernel.org
Precedence: bulk
List-ID: <kvm.vger.kernel.org>
X-Mailing-List: kvm@vger.kernel.org
X-Virus-Scanned: ClamAV using ClamSMTP

The current tap13 support summarizes testcases on a rather coarse
granularity. Which sort of defeats the purpose, especially in CI
environments, where having results for individual tests to pinpoint
problems helps a lot. Plus it can be convenient to have the plan in front
of the individual testcases.
This patch improves things by post-processing the log files, extracting
results for individual testcases, introduces test case numbers and puts the
plan in front of the test case results.
One could argue that post-processing the logs is a rather fragile approach.
Which is true - but apparently already the case, see e.g. extract_summary
in scripts/runtime.bash. Plus this is quite cheap, while a proper solution
would require to modify the kernels, so we can e.g. pass in a counter for
the test case number. And we would probably have to come up with reasonable
test case names, while this approach simply derives them from each test
output.

Before:
  $ ./run_tests.sh -t
  TAP version 13
  ok selftest-setup
  ok intercept
  ok emulator
  ok sieve
  ok sthyi
  ok skey
  ok diag10
  ok pfmf
  ok cmm
  ok vector
  ok gs # SKIP
  ok iep # SKIP
  1..12

After:
  $ ./run_tests.sh -t
  TAP version 13
  1..180
  ok 1 - cmm: privileged: Program interrupt: expected(2)
  ok 2 - cmm: invalid ORC 8: Program interrupt: expected(6)
  ok 3 - diag10: lowcore freeing: 0x0000/0x0000: Program interrupt: expected(6)
  ok 4 - diag10: lowcore freeing: 0x1000/0x1000: Program interrupt: expected(6)
  ok 5 - diag10: lowcore freeing: 0x0000/0x1000: Program interrupt: expected(6)
  ok 6 - diag10: start/end: end < start: Program interrupt: expected(6)
  ok 7 - diag10: start/end: unaligned start: Program interrupt: expected(6)
  ok 8 - diag10: start/end: unaligned end: Program interrupt: expected(6)
  ok 9 - diag10: privileged: Program interrupt: expected(2)
  ok 10 - emulator: spm/ipm: cc=0,key=f: bit 32 and 33 set to zero
  ok 11 - emulator: spm/ipm: cc=0,key=f: bit 0-31, 40-63 unchanged
  ok 12 - emulator: spm/ipm: cc=0,key=f: cc and key applied
  ok 13 - emulator: spm/ipm: cc=1,key=9: bit 32 and 33 set to zero
  [...]

Signed-off-by: Stefan Raspl <raspl@linux.ibm.com>
---
 run_tests.sh         | 10 +++++++---
 scripts/runtime.bash | 47 +++++++++++++++++++++++++++++++++++++++--------
 2 files changed, 46 insertions(+), 11 deletions(-)

diff --git a/run_tests.sh b/run_tests.sh
index 6b0af19..01acc38 100755
--- a/run_tests.sh
+++ b/run_tests.sh
@@ -84,7 +84,6 @@ function run_task()
 	if [ -z "$testname" ]; then
 		return
 	fi
-	test_number=$((test_number+1))
 
 	while (( $(jobs | wc -l) == $unittest_run_queues )); do
 		# wait for any background test to finish
@@ -110,9 +109,13 @@ mkdir $unittest_log_dir || exit 2
 echo "BUILD_HEAD=$(cat build-head)" > $unittest_log_dir/SUMMARY
 
 if [[ $tap_output == "yes" ]]; then
-    test_number=0
+    tap13_lock="`mktemp`"
+    tap13_count="`mktemp`"
+    rm "$tap13_lock"
+    echo 0 >"$tap13_count"
     echo "TAP version 13"
 fi
+
 trap "wait; exit 130" SIGINT
 for_each_unittest $config run_task
 
@@ -120,5 +123,6 @@ for_each_unittest $config run_task
 wait
 
 if [[ $tap_output == "yes" ]]; then
-    echo "1..$test_number"
+    echo "1..`cat "$tap13_count"`"
+    rm $tap13_count
 fi
diff --git a/scripts/runtime.bash b/scripts/runtime.bash
index 2f8026d..f31a80b 100644
--- a/scripts/runtime.bash
+++ b/scripts/runtime.bash
@@ -69,15 +69,44 @@ function print_result()
         fi
         return
     fi
+}
 
-    if [[ $status == "FAIL" ]]; then
-        echo "not ok $testname $reason"
-    elif [[ $status == "PASS" ]]; then
-        echo "ok $testname"
-    elif [[ $status == "SKIP" ]]; then
-        echo "ok $testname # SKIP $reason"
-    else
-        echo "not ok # TODO unknown test status"
+function get_lock () {
+    for (( ;; )); do
+        ln -s "$tap13_count" "$tap13_lock" 2>/dev/null
+        if [ $? -eq 0 ]; then
+            break
+        fi
+        sleep 1
+    done
+}
+
+function generate_tap13_output() {
+    if [[ $tap_output == "yes" ]]; then
+        get_lock
+        test_number=`cat "$tap13_count"`
+        while read -r line; do
+            name="`echo ${line% ==*} | cut -c 7-`"
+            case "${line:0:4}" in
+                PASS)
+                    (( test_number++ ))
+                    echo "ok $test_number - $name"
+                    ;;
+                FAIL)
+                    (( test_number++ ))
+                    echo "not ok $test_number - $name"
+                    ;;
+                SKIP)
+                    (( test_number++ ))
+                    echo "ok $test_number - $name # skip"
+                    ;;
+                *)
+                    continue
+                    ;;
+            esac
+        done < "$unittest_log_dir/$1.log"
+        echo $test_number>"$tap13_count"
+        rm "$tap13_lock"
     fi
 }
 
@@ -161,6 +190,8 @@ function run()
         print_result "FAIL" $testname "$summary"
     fi
 
+    generate_tap13_output "$testname"
+
     return $ret
 }
 

From patchwork Fri Feb 22 14:38:59 2019
Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: 7bit
X-Patchwork-Submitter: Stefan Raspl <raspl@linux.ibm.com>
X-Patchwork-Id: 10826201
Return-Path: <kvm-owner@kernel.org>
Received: from mail.wl.linuxfoundation.org (pdx-wl-mail.web.codeaurora.org
 [172.30.200.125])
	by pdx-korg-patchwork-2.web.codeaurora.org (Postfix) with ESMTP id 2E42D1823
	for <patchwork-kvm@patchwork.kernel.org>;
 Fri, 22 Feb 2019 14:39:11 +0000 (UTC)
Received: from mail.wl.linuxfoundation.org (localhost [127.0.0.1])
	by mail.wl.linuxfoundation.org (Postfix) with ESMTP id 18D60305E7
	for <patchwork-kvm@patchwork.kernel.org>;
 Fri, 22 Feb 2019 14:39:11 +0000 (UTC)
Received: by mail.wl.linuxfoundation.org (Postfix, from userid 486)
	id 0CAC83072C; Fri, 22 Feb 2019 14:39:11 +0000 (UTC)
X-Spam-Checker-Version: SpamAssassin 3.3.1 (2010-03-16) on
	pdx-wl-mail.web.codeaurora.org
X-Spam-Level: 
X-Spam-Status: No, score=-7.9 required=2.0 tests=BAYES_00,MAILING_LIST_MULTI,
	RCVD_IN_DNSWL_HI autolearn=ham version=3.3.1
Received: from vger.kernel.org (vger.kernel.org [209.132.180.67])
	by mail.wl.linuxfoundation.org (Postfix) with ESMTP id 9334730731
	for <patchwork-kvm@patchwork.kernel.org>;
 Fri, 22 Feb 2019 14:39:10 +0000 (UTC)
Received: (majordomo@vger.kernel.org) by vger.kernel.org via listexpand
        id S1726494AbfBVOjJ (ORCPT
        <rfc822;patchwork-kvm@patchwork.kernel.org>);
        Fri, 22 Feb 2019 09:39:09 -0500
Received: from mx0b-001b2d01.pphosted.com ([148.163.158.5]:33826 "EHLO
        mx0a-001b2d01.pphosted.com" rhost-flags-OK-OK-OK-FAIL)
        by vger.kernel.org with ESMTP id S1726278AbfBVOjJ (ORCPT
        <rfc822;kvm@vger.kernel.org>); Fri, 22 Feb 2019 09:39:09 -0500
Received: from pps.filterd (m0098416.ppops.net [127.0.0.1])
        by mx0b-001b2d01.pphosted.com (8.16.0.27/8.16.0.27) with SMTP id
 x1MEakbI149502
        for <kvm@vger.kernel.org>; Fri, 22 Feb 2019 09:39:08 -0500
Received: from e06smtp01.uk.ibm.com (e06smtp01.uk.ibm.com [195.75.94.97])
        by mx0b-001b2d01.pphosted.com with ESMTP id 2qtk0983ue-1
        (version=TLSv1.2 cipher=AES256-GCM-SHA384 bits=256 verify=NOT)
        for <kvm@vger.kernel.org>; Fri, 22 Feb 2019 09:39:07 -0500
Received: from localhost
        by e06smtp01.uk.ibm.com with IBM ESMTP SMTP Gateway: Authorized Use
 Only! Violators will be prosecuted
        for <kvm@vger.kernel.org> from <raspl@linux.ibm.com>;
        Fri, 22 Feb 2019 14:39:06 -0000
Received: from b06cxnps4076.portsmouth.uk.ibm.com (9.149.109.198)
        by e06smtp01.uk.ibm.com (192.168.101.131) with IBM ESMTP SMTP Gateway:
 Authorized Use Only! Violators will be prosecuted;
        (version=TLSv1/SSLv3 cipher=AES256-GCM-SHA384 bits=256/256)
        Fri, 22 Feb 2019 14:39:02 -0000
Received: from d06av22.portsmouth.uk.ibm.com (d06av22.portsmouth.uk.ibm.com
 [9.149.105.58])
        by b06cxnps4076.portsmouth.uk.ibm.com (8.14.9/8.14.9/NCO v10.0) with
 ESMTP id x1MEd1MJ10879134
        (version=TLSv1/SSLv3 cipher=DHE-RSA-AES256-GCM-SHA384 bits=256
 verify=OK);
        Fri, 22 Feb 2019 14:39:01 GMT
Received: from d06av22.portsmouth.uk.ibm.com (unknown [127.0.0.1])
        by IMSVA (Postfix) with ESMTP id 756D04C046;
        Fri, 22 Feb 2019 14:39:01 +0000 (GMT)
Received: from d06av22.portsmouth.uk.ibm.com (unknown [127.0.0.1])
        by IMSVA (Postfix) with ESMTP id 613F34C040;
        Fri, 22 Feb 2019 14:39:01 +0000 (GMT)
Received: from tuxmaker.boeblingen.de.ibm.com (unknown [9.152.85.9])
        by d06av22.portsmouth.uk.ibm.com (Postfix) with ESMTPS;
        Fri, 22 Feb 2019 14:39:01 +0000 (GMT)
Received: from tuxmaker.boeblingen.de.ibm.com (localhost [127.0.0.1])
        by tuxmaker.boeblingen.de.ibm.com (Postfix) with ESMTP id 0C9EE20F5F7;
        Fri, 22 Feb 2019 15:39:01 +0100 (CET)
From: Stefan Raspl <raspl@linux.ibm.com>
To: kvm@vger.kernel.org
Cc: drjones@redhat.com, pbonzini@redhat.com, sergeyb@bronevichok.ru,
        heiko.carstens@de.ibm.com
Subject: [kvm-unit-tests PATCH v2 2/2] run_tests: fix command line options
 handling
Date: Fri, 22 Feb 2019 15:38:59 +0100
X-Mailer: git-send-email 2.16.4
In-Reply-To: <20190222143859.15382-1-raspl@linux.ibm.com>
References: <20190222143859.15382-1-raspl@linux.ibm.com>
X-TM-AS-GCONF: 00
x-cbid: 19022214-4275-0000-0000-00000312B4EB
X-IBM-AV-DETECTION: SAVI=unused REMOTE=unused XFE=unused
x-cbparentid: 19022214-4276-0000-0000-00003820ECBB
Message-Id: <20190222143859.15382-3-raspl@linux.ibm.com>
X-Proofpoint-Virus-Version: vendor=fsecure engine=2.50.10434:,,
 definitions=2019-02-22_10:,,
 signatures=0
X-Proofpoint-Spam-Details: rule=outbound_notspam policy=outbound score=0
 priorityscore=1501
 malwarescore=0 suspectscore=1 phishscore=0 bulkscore=0 spamscore=0
 clxscore=1015 lowpriorityscore=0 mlxscore=0 impostorscore=0
 mlxlogscore=999 adultscore=0 classifier=spam adjust=0 reason=mlx
 scancount=1 engine=8.0.1-1810050000 definitions=main-1902220104
Sender: kvm-owner@vger.kernel.org
Precedence: bulk
List-ID: <kvm.vger.kernel.org>
X-Mailing-List: kvm@vger.kernel.org
X-Virus-Scanned: ClamAV using ClamSMTP

When specifying desired tests on the command line, options are ignored
unless specified prior to any desired tests.
E.g.

  $ ./run_tests.sh sthyi -t emulator -j5
  PASS emulator (102 tests)
  PASS sthyi (19 tests)

Note that the output is not in TAP13 format, and the tests are not executed
in parallel.
This patch fixes the issue by switching to getopt instead. Might not be the
optimum solution, but threw in long command line opts as a bonus.

Signed-off-by: Stefan Raspl <raspl@linux.ibm.com>
---
 run_tests.sh | 47 +++++++++++++++++++++++++++--------------------
 1 file changed, 27 insertions(+), 20 deletions(-)

diff --git a/run_tests.sh b/run_tests.sh
index 01acc38..bb13fbf 100755
--- a/run_tests.sh
+++ b/run_tests.sh
@@ -17,13 +17,13 @@ cat <<EOF
 
 Usage: $0 [-h] [-v] [-a] [-g group] [-j NUM-TASKS] [-t]
 
-    -h: Output this help text
-    -v: Enables verbose mode
-    -a: Run all tests, including those flagged as 'nodefault'
-        and those guarded by errata.
-    -g: Only execute tests in the given group
-    -j: Execute tests in parallel
-    -t: Output test results in TAP format
+    -h, --help      Output this help text
+    -v, --verbose   Enables verbose mode
+    -a, --all       Run all tests, including those flagged as 'nodefault'
+                    and those guarded by errata.
+    -g, --group     Only execute tests in the given group
+    -j, --parallel  Execute tests in parallel
+    -t, --tap13     Output test results in TAP format
 
 Set the environment variable QEMU=/path/to/qemu-system-ARCH to
 specify the appropriate qemu binary for ARCH-run.
@@ -34,39 +34,46 @@ EOF
 RUNTIME_arch_run="./$TEST_DIR/run"
 source scripts/runtime.bash
 
-while getopts "ag:htj:v" opt; do
-    case $opt in
-        a)
+only_tests=""
+args=`getopt -u -o ag:htj:v -l all,group:,help,tap13,parallel:,verbose -- $*`
+[ $? -ne 0 ] && exit 2;
+set -- $args;
+while [ $# -gt 0 ]; do
+    case "$1" in
+        -a | --all)
             run_all_tests="yes"
             export ERRATA_FORCE=y
             ;;
-        g)
-            only_group=$OPTARG
+        -g | --group)
+            shift
+            only_group=$1
             ;;
-        h)
+        -h | --help)
             usage
             exit
             ;;
-        j)
-            unittest_run_queues=$OPTARG
+        -j | --parallel)
+            shift
+            unittest_run_queues=$1
             if (( $unittest_run_queues <= 0 )); then
                 echo "Invalid -j option: $unittest_run_queues"
                 exit 2
             fi
             ;;
-        v)
+        -v | --verbose)
             verbose="yes"
             ;;
-        t)
+        -t | --tap13)
             tap_output="yes"
             ;;
+        --)
+            ;;
         *)
-            exit 2
+            only_tests="$only_tests $1"
             ;;
     esac
+    shift
 done
-shift $((OPTIND - 1))
-only_tests="$*"
 
 # RUNTIME_log_file will be configured later
 RUNTIME_log_stderr () { cat >> $RUNTIME_log_file; }
